Adding jitpack support

quintesse opened the pull request:
You can now add URLs to repos from github.com, gitlab.com and bitbucket.org.

You can use links to the root of the master branch, to any tagged release or to a specific commit and they will be turned into GAV coordinates for JitPack. If the JitPack repo wasn't added by the developer it will be added automatically.

maxandersen commented on src/main/java/dk/xam/jbang/Script.java:
```diff
@@ -221,11 +237,102 @@ public String resolveClassPath(boolean offline) {
+		return ref;
+	}
+
+	static String githubTreeUrlToGAV(String ref) {
```
all this magic git ref to GAV feels like it deserves its own class/utility

quintesse commented:
So how about this new version?

quintesse commented:
Updated again, this is the fully working version that also automatically adds the JitPack repo if it wasn't added explicitly. I added an example, "urldep.java", to test it.

maxandersen commented:
looks great @quintesse  - could you add a blurp to the readme.adoc around `//DEPS` ?

quintesse commented:
@maxandersen added support for the `#` options we talked about. Refactored the code a bit to make things somewhat cleaner and easier to implement because with those new options things were getting a little crazy. And added some lines to the README too.

maxandersen commented:
building and then doing `jbang examples/urldep.java` gives me::

```
java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractList.add(AbstractList.java:153)
	at java.base/java.util.AbstractList.add(AbstractList.java:111)
	at dk.xam.jbang.DependencyUtil.resolveDependencies(DependencyUtil.java:65)
	at dk.xam.jbang.Script.resolveClassPath(Script.java:149)
	at dk.xam.jbang.Main.build(Main.java:365)
	at dk.xam.jbang.Main.doCall(Main.java:290)
	at dk.xam.jbang.Main.call(Main.java:340)
	at dk.xam.jbang.Main.call(Main.java:74)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
	at picocli.CommandLine.execute(CommandLine.java:1904)
	at dk.xam.jbang.Main.main(Main.java:171)
```

looking to see whats going on...

maxandersen commented:
okey so something funky is going on with unmodifiable lists..fixing.

maxandersen commented:
fixed the immutable list...but now:

```
[jbang] Resolving dependencies...
[jbang]     Resolving com.github.kohsuke:github-api:github-api-1.107...Apr 16, 2020 11:26:08 PM org.jboss.shrinkwrap.resolver.impl.maven.logging.LogTransferListener transferFailed
WARNING: Failed downloading com/github/kohsuke/github-api/github-api-1.107/github-api-github-api-1.107.pom from https://jcenter.bintray.com/. Reason:
org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.github.kohsuke:github-api:pom:github-api-1.107 in jcenter (https://jcenter.bintray.com/)
Apr 16, 2020 11:26:10 PM org.jboss.shrinkwrap.resolver.impl.maven.logging.LogTransferListener transferFailed
WARNING: Failed downloading com/github/kohsuke/github-api/github-api-1.107/github-api-github-api-1.107.jar from https://jcenter.bintray.com/. Reason:
org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.github.kohsuke:github-api:jar:github-api-1.107 in jcenter (https://jcenter.bintray.com/)
Done
[jbang] Dependencies resolved
[jbang] Building jar...
Picked up JAVA_TOOL_OPTIONS:
Picked up JAVA_TOOL_OPTIONS:
Exception in thread "main" org.kohsuke.github.HttpException: Server returned HTTP response code: 200, message: 'null' for URL: https://api.github.com/repos/maxandersen/jbang
	at org.kohsuke.github.GitHubClient.interpretApiError(GitHubClient.java:441)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:363)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:307)
	at org.kohsuke.github.Requester.fetch(Requester.java:68)
	at org.kohsuke.github.GitHub.getRepository(GitHub.java:522)
	at urldep.main(urldep.java:15)
Caused by: com.fasterxml.jackson.core.JsonParseException: Illegal character ((CTRL-CHAR, code 31)): only regular white space (\r, \n, \t) is allowed between tokens
 at [Source: (String)"�Xao�8�+������EZ�tժ����ޭ����G�������w���R �{yόg���;���o�^7�1��g݇��6��"�r��?�ZG�j���m�0��~[��b1MV.g4[��bZ?K��b�4�����9����S�Y�+7S�W�k��&�Z�����x�\����?��h�c�0�a)U�B	P$��zL�{���������Hf�efɔ�b���t
!7�8�|�5�Z[W$<[��Xi�a;�\�7K�@%��K�q��o�U� �:�sI�e�W�t�xn���,�
                                                            2�4�;�E���f`����P%�(�Z�(1����O� 4����x��:7lJ�Ԇe��=��+}�	��5Gg�3��N:�Nn�#睯���\��N��z�K'R��\�bDV�mGY�̫
                             �>F�
S��֓�"K��uhD�Y:��y)�ω:�(I���
�h�)���"�"[truncated 737 chars]; line: 1, column: 2]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1840)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:712)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._throwInvalidSpace(ParserMinimalBase.java:690)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._skipWSOrEnd(ReaderBasedJsonParser.java:2373)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:672)
	at com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:4340)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4189)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3205)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3173)
	at org.kohsuke.github.GitHubResponse.parseBody(GitHubResponse.java:80)
	at org.kohsuke.github.Requester.lambda$fetch$1(Requester.java:68)
	at org.kohsuke.github.GitHubClient.createResponse(GitHubClient.java:401)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:355)
```

veyr weird

maxandersen commented:
okey so something is wrong with that dep - using `//DEPS https://github.com/kohsuke/github-api/tree/github-api-1.101` makes it work.

maxandersen commented:
added a few commits to make `#:` and `#:snapshot` working.

quintesse commented:
Cool! Thanks :-)
